online-banking-cz/src/main/java/com/adyen/checkout/onlinebankingcz/PdfOpener.kt

```diff
@@ -0,0 +1,98 @@
+/*
+ * Copyright (c) 2022 Adyen N.V.
+ *
+ * This file is open source and available under the MIT license. See the LICENSE file for more info.
+ *
+ * Created by atef on 13/9/2022.
+ */
+
+package com.adyen.checkout.onlinebankingcz
+
+import android.content.ActivityNotFoundException
+import android.content.Context
+import android.content.Intent
+import android.net.Uri
+import android.os.Build
+import androidx.browser.customtabs.CustomTabColorSchemeParams
+import androidx.browser.customtabs.CustomTabsIntent
+import com.adyen.checkout.components.ui.util.ThemeUtil
+import com.adyen.checkout.core.log.LogUtil
+import com.adyen.checkout.core.log.Logger
+
+internal class PdfOpener {
+
+    @Throws(IllegalStateException::class)
+    fun open(context: Context, url: String) {
+        val uri = Uri.parse(url)
+        if (open(context, uri)) return
```

Comment:
As I was saying yesterday if `openInExternalApp` fails then `openInBrowser` browser will be called. It is better to first fallback to `openInCustomTab` before opening the browser, so there is one app less in this flow. To achieve this with this structure, this method:
```Kotlin
private fun open(context: Context, uri: Uri): Boolean {
    return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) openInExternalApp(context, uri)
    // because custom tabs pdf viewer is working before api 30 on chrome
    else openInCustomTab(context, uri)
}
```
Has to become something like this:
```Kotlin
private fun open(context: Context, uri: Uri): Boolean {
    return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
        if (openInExternalApp(context, uri)) true
        else openInCustomTab(context, uri)
    }
    else openInCustomTab(context, uri)
}
```
Does it become more clear now why I wrote that if statement like that in the example I gave before? In this way the flow is in multiple methods and nested if statements, which is more difficult to follow. In my example the whole flow is in just this one if statement (I do agree it's not the prettiest, but it's just 3 lines you have to follow)

Reply:
yeah I guess that's why I added an extra if statement for `openInCustomTab` in the first place like this
```
    @Throws(IllegalStateException::class)
    fun open(context: Context, url: String) {
        val uri = Uri.parse(url)
        if (open(context, uri)) return
        if (openInCustomTab(context, uri)) return
        if (openInBrowser(context, uri)) return

        Logger.e(TAG, "Couldn't open pdf with url: $uri")

        throw IllegalStateException("Couldn't open pdf with url: $uri")
    }
```
I guess now I am gonna go with you're suggestion because `openInCustomTab(context, uri)` will be called twice for api level 30